### PR TITLE
Handling crash due to configuration change / application minimization by using Saved State module for ViewModel

### PR DESCRIPTION
**Pull request by kashyapj2793**
Addressing #147

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt [hunk 1]
```diff
@@ -61,5 +72,12 @@ class QuestionnaireFragment(private val questionnaire: Questionnaire) : Fragment
     companion object {
         const val QUESTIONNAIRE_RESPONSE_REQUEST_KEY = "questionnaire-response-request-key"
         const val QUESTIONNAIRE_RESPONSE_BUNDLE_KEY = "questionnaire-response-bundle-key"
+        @JvmStatic
+        fun newInstance(questionnaire: Questionnaire) = QuestionnaireFragment().apply {
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:76]: actually having thought about this i'm not so sure this static function is so useful. see comment below in the activity class.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt [hunk 2]
```diff
@@ -42,6 +49,10 @@ class QuestionnaireFragment(private val questionnaire: Questionnaire) : Fragment
     ) = inflater.inflate(R.layout.questionnaire_fragment, container, false)
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+
+        val viewModel: QuestionnaireViewModel by viewModels {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:53]: why move this inside the `onViewCreate` function? I would think the view model should still be a `private val` of the class.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt [hunk 3]
```diff
@@ -30,9 +31,15 @@ import androidx.recyclerview.widget.RecyclerView
...
-class QuestionnaireFragment(private val questionnaire: Questionnaire) : Fragment() {
-    private val viewModel: QuestionnaireViewModel by viewModels {
-        QuestionnaireViewModelFactory(questionnaire)
+class QuestionnaireFragment() : Fragment() {
+
+    private lateinit var questionnaire: Questionnaire
+
+    override fun onAttach(context: Context) {
+        super.onAttach(context)
+        val jsonParser = FhirContext.forR4().newJsonParser()
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:40]: i think reading the questionnaire content should be in `onCreate`, not `onAttach`.

see the example here: https://developer.android.com/guide/fragments/create#kotlin

also use `requireArguments`:

```
class QuestionnaireFragment : Fragment() {
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val questionnaireJson = requireArguments().getString("questionnaireJson")
        ...
    }
}
```

**Comment by jingtang10** [comment]
Thanks so much for fixing this Kashyap!

Sorry for the delay - I wasn't sure if this was ready for review because it was a draft. Please feel free to convert drafts to PRs whenever they're ready to be reviewed.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (kashyapj2793) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt]: Yes according to @yigit  comments in PR #144 I switched to passing the SavedStateHandle to the ViewModel as described in [https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate](url) and passing the quesitonnaire as a Bundle to the QuestionnaireFragment.arguments

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt (hunk 2 above)
Comment (kashyapj2793) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt]: I have made it a private val of the class again

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt [hunk 4]
```diff
@@ -28,11 +28,17 @@ import androidx.fragment.app.setFragmentResult
...
 
-class QuestionnaireFragment(private val questionnaire: Questionnaire) : Fragment() {
     private val viewModel: QuestionnaireViewModel by viewModels {
-        QuestionnaireViewModelFactory(questionnaire)
+        QuestionnaireViewModelFactory(this, requireArguments())
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        if (arguments == null) throw IllegalStateException("Pass arg \'questionnaire\' to fragment")
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:41]: is this needed? wouldn't `requireArguments()` above already do this?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt [hunk 5]
```diff
@@ -42,6 +48,7 @@ class QuestionnaireFragment(private val questionnaire: Questionnaire) : Fragment
     ) = inflater.inflate(R.layout.questionnaire_fragment, container, false)
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:51]: nit: remove extra line

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 6]
```diff
@@ -16,23 +16,31 @@
 
 package com.google.android.fhir.datacapture
 
+import android.os.Bundle
+import androidx.lifecycle.AbstractSavedStateViewModelFactory
+import androidx.lifecycle.SavedStateHandle
 import androidx.lifecycle.ViewModel
-import androidx.lifecycle.ViewModelProvider
+import androidx.savedstate.SavedStateRegistryOwner
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:23]: unused imports?

File: datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/MainActivity.kt [hunk 7]
```diff
@@ -30,32 +31,35 @@ class MainActivity : AppCompatActivity() {
...
         supportFragmentManager.setFragmentResultListener(
             QuestionnaireFragment.QUESTIONNAIRE_RESPONSE_REQUEST_KEY,
             this,
             object : FragmentResultListener {
                 override fun onFragmentResult(requestKey: String, result: Bundle) {
-                    val dialogFragment = QuestionnaireResponseDialogFragment(
-                        result.getString(QuestionnaireFragment.QUESTIONNAIRE_RESPONSE_BUNDLE_KEY)!!
-                    )
+                    val dialogFragment = QuestionnaireResponseDialogFragment()
+                    dialogFragment.arguments = bundleOf(Pair("contents", result.getString
```
Comment (jingtang10) [inline datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/MainActivity.kt:47]: can you use the kotlin syntax:

`bundleOf(x to y)`

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/to.html

File: datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/QuestionnaireResponseDialogFragment.kt [hunk 8]
```diff
@@ -22,8 +22,9 @@ import android.os.Bundle
 import android.widget.TextView
 import androidx.fragment.app.DialogFragment
 
-class QuestionnaireResponseDialogFragment(val contents: String) : DialogFragment() {
+class QuestionnaireResponseDialogFragment() : DialogFragment() {
     override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        val contents = requireArguments().getString("contents")
```
Comment (jingtang10) [inline datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/QuestionnaireResponseDialogFragment.kt:27]: the key "contents" needs to be  a const value extracted somewhere -- error prone to be hardcoded like that.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt (hunk 4 above)
Comment (kashyapj2793) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:41]: Yes it does, but I thought we should be giving a more appropriate error message.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 6 above)
Comment (kashyapj2793) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:23]: It is used in the constructor of the QuestionnaireViewModelFactory

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt]: fantastic! thanks!

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt (hunk 4 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:41]: Well actually this message is too specific --  the condition only checks if arguments exists but not whether `questionnaire` is specified.

I personally think the code is simpler without this and the default message for requireArguments ("Fragment " + this + " does not have any arguments.") is good enough.

Comment (kashyapj2793) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:41]: ok sure Ill remove this check 👍🏼

**Review by jingtang10 — APPROVED** [review]

File: datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/QuestionnaireResponseDialogFragment.kt (hunk 8 above)
Comment (kashyapj2793) [reply datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/QuestionnaireResponseDialogFragment.kt]: done, also moved "questionnaire" to a constant value

File: datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/MainActivity.kt (hunk 7 above)
Comment (kashyapj2793) [reply datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/MainActivity.kt]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt (hunk 4 above)
Comment (kashyapj2793) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt]: done

**Pull request merged**